### PR TITLE
Multidrive support

### PR DESCRIPTION
**Pull request by joehand**
I think I finished this before the break but need to check it over. ~Depends on https://github.com/joehand/dat-folder-archive/pull/2.~ (Implemented this directly in here instead of in dat-folder-archive).

I changed the API slightly around the network to make it easier to use an external swarm module. But everything is still backwards compatible.

## TODO

* Not doing: ~Release MD support in dat-folder-archiver~
* Done: ~Add MD test here~
* Done: ~pfrazee: updating the usage to simplify it, so you can just do `Dat(drive, function(err, dat) {}`~
* [Done](https://github.com/datproject/dat-node/pull/65#issuecomment-272560931)  ~Ensure compatibility with dat cli (to do this we need to put the archive key into the `.dat` db)~
- [x] Review & update docs
- [x] Check compatibility with dat-next, old CLI.
- [x] Merge!

### CODE REVIEW

**Comment by juliangruber** [comment]
Why does this depend on https://github.com/joehand/dat-folder-archive/pull/2/files? Wouldn't it be the other way around?

**Comment by joehand** [comment]
> Why does this depend on https://github.com/joehand/dat-folder-archive/pull/2/files, wouldn't it be the other way around?

I use dat-folder-archive to [create the archive](https://github.com/datproject/dat-node/blob/master/lib/init-archive.js#L27). So I need to pass through `opts.drive` here and use the drive in `dat-folder-archive` to make the archive.

**Comment by juliangruber** [comment]
ah, oh I think I was confused by the module's name and should have written it's readme. I thought dat-folder-archive was a part of dat-folder.com

**Comment by joehand** [comment]
> ah, oh I think I was confused by the module's name and should have written it's readme. I thought dat-folder-archive was a part of dat-folder.com

Ya =(. I named that before we named dat folder.

File: dat.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  target = target.length ? target : self.path
+
+  self.importer = importFiles(self.archive, target, opts, function (err) {
+    if (err || self.archive.live) return cb(err)
+    // Sets self.key for snapshot
+    self.archive.finalize(function (err) {
+      if (err) return cb(err)
+      self.key = self.archive.key
+      // TODO: need to get snapshot key back in db, better way?
+      if (self.db) self.db.put('!dat!key', self.archive.key.toString('hex'), cb)
```
Comment (karissa) [inline dat.js]: Should we use `encoding.str(self.archive.key)` here?

File: dat.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  opts = opts || {}
+  assert.ok(opts.db || this.db, 'Dat needs database to track stats')
+  this.stats = stats(this.archive, opts.db || this.db)
+  return this.stats
+}
+
+Dat.prototype.importFiles = function (target, opts, cb) {
+  if (typeof target !== 'string') return this.importFiles('', target, opts)
+  if (typeof opts === 'function') return this.importFiles(target, {}, opts)
+  if (!this.archive.owner) return cb(new Error('Must be archive owner to import files.'))
```
Comment (karissa) [inline dat.js]: Will `archive.owner` only be known if the archive is opened?

File: tests/misc.js [hunk 3]
```diff
@@ -105,6 +106,27 @@ test('custom db option', function (t) {
...
 })
 
+test('custom drive option', function (t) {
+  rimraf.sync(path.join(shareFolder, '.dat'))
+  var drive = hyperdrive(memdb())
+  Dat(shareFolder, {drive: drive}, function (err, dat) {
+    t.error(err)
+    dat.archive.open(function () {
+      // Need open otherwise get DeferredLevelDOWN
+      t.ok(dat.db.db instanceof require('memdown'), 'db is memdown')
```
Comment (karissa) [inline tests/misc.js]: Can we put the require statement at the top of the file?

File: lib/network.js [hunk 4]
```diff
@@ -6,14 +6,6 @@ module.exports = function (archive, opts) {
...
-    swarm: swarm,
-    options: swarm.options || opts
-  }
-
-  network.__defineGetter__('connected', function () {
-    return swarm.connected
-  })
-
-  return network
+  swarm.options = swarm.options || opts
```
Comment (karissa) [inline lib/network.js]: Is this line necessary? Confused about if `opts` above becomes `swarm.options` under the hood

File: dat.js [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  assert.ok(opts.dir, 'opts.directory required')
+
+  this.path = path.resolve(opts.dir)
+  this.options = opts
+
+  this.archive = archive
+  this.db = db
+  this.key = archive.key // only resumed/owned archives will have keys here
+  this.live = archive.live
+  this.owner = archive.owner
```
Comment (karissa) [inline dat.js]: Do I remember it right that `archive.owner` is only populated after `archive.open`?

Comment (joehand) [inline dat.js]: Yep! I force the archive open in [init-archive](https://github.com/datproject/dat-node/blob/multidrive-support/lib/init-archive.js#L33) and set owner to false if the archive is being downloaded. As I say in my comments there, its a bit hacky.

File: dat.js (hunk 2 above)
Comment (joehand) [inline dat.js]: Ya. It'll be true if the archive is local. For future multi-writer, we can set it after a [downloaded archive](https://github.com/datproject/dat-node/blob/multidrive-support/dat.js#L42) is opened.

File: dat.js (hunk 5 above)
Comment (karissa) [inline dat.js]: OK cool!

File: lib/network.js (hunk 4 above)
Comment (joehand) [inline lib/network.js]: Darn. I can't remember why I put this in there.

Looks like hyperdiscovery doesn't set `swarm.options` and discovery-swarm sets `opts` to `swarm._options`. I probably just saw it was null in some situation so used this. But we should probably expose `swarm._options` directly here instead.

**Comment by joehand** [comment]
[https://github.com/datproject/dat-node/pull/65/commits/baafacba271cb1460b81c77c6efd606eedf6cd58] was a pretty big change working to ensure compatibility across multidrive or other apps. 

tldr; previously I was saving the archive key in the `.dat` folder to make it easy to resume. Now I use `core.list()` to get all the keys in a database and make an educated guess about which one is the metadata and if to resume.

Also:

* Allow drive as first argument (though I still need to work out when/if dir is required)
* Automatically open more archives (anything that is local will be opened)
* Better resuming cross compatibility by not storing the key in the dir but determining existing archive from the keys in the core.

**Comment by joehand** [comment]
Code is ready. Just docs and making sure this works with CLIs left!
